flasksrc/session.py

```diff
@@ -168,23 +173,76 @@ def getManager():
     @staticmethod
     def getContainer(session):
         if session in State.containers:
+            State.last_used[session] = time.time()
             return State.manager.running_containers[State.containers[session]]
         else:
             return None
 
     @staticmethod
     def setContainer(session, container):
         State.containers[session] = container
+        State.last_used[session] = time.time()
 
     @staticmethod
     def delContainer(session):
         res = False
         if session in State.containers:
             res = State.manager.killContainer(State.containers[session])
-            if res:
-                del State.containers[session]
+            if not res:
+                print(f'Failed to terminate container for session {session}')
+                # Enqueue to hopefully delete later
+                State.reaper_queue(session, None)
+
+            # Invalidate container references for that session
+            State.containers.pop(session, None)
+            State.last_used.pop(session, None)
+        else:
+            print(f'Session not tracked in State.containers: {session}')
         return res
 
+    @staticmethod
+    def expireContainer():
+        '''This method iterates through all current sessions and times them out
+        if they are stale for longer than 15 minutes.
+
+        It is intended to be called every time there is a call to this API.
+        This is another stopgap measure to conserve resources until we offload
+        state from the backend.
+        '''
+        # First pass, put expired containers into reaper queue
+        candidates = list(State.last_used.keys()).copy()
+        for session in candidates:
+            now = time.time()
+            # Guard against modifying the dict mid-iteration
+            if session in State.last_used:
+                a_time = State.last_used[session]
+                if now - a_time > SESSION_TIMEOUT:
+                    State.reaper_queue.append(State.containers[session])
+                    State.containers.pop(session, None)
+                    State.last_used.pop(session, None)
+
+        # Iterate through and kill expired containers, only do 2 at a time to
+        # reduce lateny if many containers expire
+        tmp_queue = State.reaper_queue.copy()
+        killed = 0
+        for container_name in tmp_queue:
+            res = State.manager.killContainer(container_name)
+            if res:
+                State.reaper_queue.remove(container_name)
+                killed += 1
+            if killed > 1:
+                break
```

Comment:
Can we print that we killed a container successfully? That way we can verify this in the logs as well as checking docker ps

Reply:
Added